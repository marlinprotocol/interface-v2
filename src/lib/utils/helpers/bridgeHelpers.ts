import {
	DEFAULT_PRECISION,
	MPOND_PRECISIONS,
	POND_PRECISIONS
} from '$lib/utils/constants/constants';

export const getAmountPrecision = (token: 'pond' | 'mPond') => {
	switch (token) {
		case 'pond':
			return POND_PRECISIONS;
		case 'mPond':
			return MPOND_PRECISIONS;
		default:
			return DEFAULT_PRECISION;
	}
};
