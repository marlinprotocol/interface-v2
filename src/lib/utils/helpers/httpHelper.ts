import { addToast } from '$lib/data-stores/v2/toastStore';

export async function fetchHttpData(url: string, options: RequestInit) {
	return fetch(url, options)
		.then((res) => res.json())
		.then((res) => {
			return res;
		})
		.catch((error) => {
			addToast({
				message: {
					title: 'HTTP Error',
					description: 'Error fetching data. Please try again later.'
				},
				variant: 'error'
			});
			console.error('error from the fetch call');
			throw new Error(error);
		});
}
