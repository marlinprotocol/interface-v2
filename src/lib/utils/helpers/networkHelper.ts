import type { EIP1193Provider } from '@web3-onboard/core';
import { environment } from '$lib/data-stores/environment';
import onboard from '$lib/controllers/web3OnboardController';
import { setWalletAndChainStores } from '$lib/controllers/walletController';

/**
 * Checks if current chain is supported by the app or not
 * @param chainId
 * @returns if chain is supported or not
 */
export function isValidChain(chainId: number): boolean {
	// check if environment.valid_chains has chainId as a key in it
	return Object.keys(environment.valid_chains).includes(chainId.toString());
}

export async function switchChain(chainId: number, provider: EIP1193Provider) {
	if (isValidChain(chainId)) {
		const success = await onboard.setChain({ chainId: chainId });
		if (success) {
			console.log('setting wallet and chain stores due to chain switch');
			setWalletAndChainStores(provider);
		}
	}
}
