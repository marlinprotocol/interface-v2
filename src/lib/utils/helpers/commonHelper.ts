import { checkIfSignerExistsInSubgraph } from '$lib/controllers/subgraphController';
import { shortenText } from '$lib/utils/helpers/conversionHelper';
import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function copyTextToClipboard(text: string) {
	if (navigator.clipboard) {
		navigator.clipboard.writeText(text);
	}
}

export function getCurrentEpochCycle(epochStartTime: number, epochLength: number): number {
	if (epochLength === 0) return 0;

	const currentEpoch = new Date().getTime() / 1000;

	//if epoch start time is in future, then epoch cycle is 0
	if (currentEpoch <= epochStartTime) {
		return 0;
	}
	const epochCycle = Math.floor((currentEpoch - epochStartTime) / epochLength) + 1;
	return epochCycle;
}

// TODO: these can be moved to a seperate modal helper file
/**
 * closes modal based on the id that is passed in <label for={modalId}> by setting the respective checkbox to true
 * @param modalId
 */
export function openModal(modalId: string) {
	const modalCheckboxInput = document.getElementById(modalId) as HTMLInputElement;
	modalCheckboxInput.checked = true;
}

/**
 * closes modal based on the id that is passed in <label for={modalId}> by setting the respective checkbox to false
 * @param modalId
 */
export function closeModal(modalId: string) {
	const modalCheckboxInput = document.getElementById(modalId) as HTMLInputElement;
	modalCheckboxInput.checked = false;
}

/**
 * checks and returns if the amount is valid or not
 */
export function isInputAmountValid(amount: string): boolean {
	if (!amount) return true;

	const containsSpecialCharsOrAlphabets = /[^\d.]/.test(amount);
	if (containsSpecialCharsOrAlphabets) return false;

	const containsMoreThanOneDecimal = (amount.match(/\./g) || []).length > 1;
	if (containsMoreThanOneDecimal) return false;

	const integerPart = amount.split('.')[0];
	const decimalPart = amount.split('.')[1];

	if (integerPart.length > 50) return false;
	if (decimalPart && decimalPart.length > 18) return false;
	if (!decimalPart && BigInt(integerPart) === 0n) return false;
	if (decimalPart && BigInt(integerPart) === 0n && BigInt(decimalPart) === 0n) return false;

	return true;
}

/**
 * Returns error message if the amount is not valid
 */
export function inputAmountInValidMessage(amount: string): string {
	const isValid = isInputAmountValid(amount);
	if (isValid) return '';
	let message = '';

	// TODO: there should be a better way to create this message since we are comparing the same thing twice once in the isAmountValid function and then here again
	if (amount === '0') {
		message = 'Amount should be greater than 0';
	} else if (amount.split('.')[0].length > 50) {
		message = 'Amount is too big.';
	} else if (amount.split('.')[1] && amount.split('.')[1].length > 18) {
		message = 'Amount can have a maximum of 18 decimals only';
	} else if (!/^\d+(\.\d{0,1})?$/.test(amount)) {
		message = 'Amount has invalid characters';
	} else if (!isValid) {
		message = 'Amount is invalid';
	}

	return message;
}

// TODO: can be removed and use shortenText function
/**
 * returns the minified address
 * @param address
 * @param first
 * @param last
 * @example minifyAddress('0x1234567890123456789012345678901234567890') => 0x123456...1234
 * @example minifyAddress('0x1234567890123456789012345678901234567890', 3, 2) => 0x123...34
 */

export function minifyAddress(
	address = '0x0000000000000000000000000000000000000000',
	first = 6,
	last = 4
): string {
	if (address === '' || first < 0 || last < 0) return '';

	const textWithoutPrefix = address.slice(2);
	if (textWithoutPrefix.length <= 2) return '';

	const shortenedText = address.slice(0, 2) + shortenText(textWithoutPrefix, first, last);
	return shortenedText;
}

/**
 * checks and returns an array of booleans signifying all the validations that the address has passed
 */
export function isAddressValid(address: string): boolean {
	const validCharacters = /^0x[a-fA-F0-9]{40}$/.test(address);

	return validCharacters;
}

/**
 * capitalizes the first letter of the string
 * @param string
 * @example capitalizeFirstLetter('hello') => 'Hello'
 */
export function capitalizeFirstLetter(string: string) {
	return string.charAt(0).toUpperCase() + string.slice(1);
}

/**
 * returns true is the url is valid
 * @param string
 * @returns
 * @example checkValidURL('https://google.com') => true
 * @example checkValidURL('http://3.108.237.212:8080') => true
 * @example checkValidURL('http://example.com/path') => true
 * @example checkValidURL('http://example.com/') => false as it has an ending slash at the end, as we send it as a query param to the backend which doesn't accept it (routing issue)
 * @example checkValidURL('example.com') => false as it has no http:// or https:// at the start
 */
export function checkValidURL(url: string) {
	const ipv4Regex = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/;
	const ipv6Regex = /^(?:[a-fA-F0-9]{1,4}:){7}[a-fA-F0-9]{1,4}$/;
	const urlRegex = /^(http|https):\/\/[\w.-]+(:\d+)?(\/\S*)?[^/]$/;

	const hasValidProtocolAndHostname = urlRegex.test(url);
	const hasNoTrailingSlashOrTrailingColon = !url.endsWith('/') && !url.endsWith(':');

	let isValid = hasValidProtocolAndHostname && hasNoTrailingSlashOrTrailingColon;

	if (!isValid) {
		return false;
	}

	const ipv4Match = ipv4Regex.test(url);
	const ipv6Match = ipv6Regex.test(url);

	if (ipv4Match || ipv6Match) {
		const urlWithoutProtocol = url.split('://')[1];
		const urlWithoutProtocolAndPath = urlWithoutProtocol.split('/')[0];
		const hostnameOrIP = urlWithoutProtocolAndPath.split(':')[0];

		if (ipv4Match) {
			const groups = hostnameOrIP.split('.');
			isValid =
				groups.length === 4 &&
				groups.every((group) => parseInt(group) >= 0 && parseInt(group) <= 255);
		} else {
			isValid = true;
		}
		return isValid;
	}

	return true;
}

/**
 * returns url appended with http:// if it doesn't have it and removes traling slashes if it has any
 * @param url
 * @example sanitizeUrl('google.com') => 'http://google.com'
 * @example sanitizeUrl('google.com/') => 'http://google.com'
 * @example sanitizeUrl('https://google.com') => 'https://google.com'
 */
export function sanitizeUrl(url: string) {
	let sanitizedUrl = url;
	if (url === '') {
		return sanitizedUrl;
	}
	if (!url.includes('://')) {
		sanitizedUrl = 'http://' + url;
	}
	if (url.endsWith('/')) {
		sanitizedUrl = sanitizedUrl.replace(/\/+$/, '');
	}
	return sanitizedUrl;
}

/**
 * returns the url for the txnHash
 * @param blockExplorerUrl (should come from $chainConfigStore.blockExplorerUrl)
 * @param txnHash
 * @example getTxnUrl('https://blockexplorer.com', '0x1234567890123456789012345678901234567890') => 'https://explorer.com/tx/0x1234567890123456789012345678901234567890'
 */
export const getTxnUrl = (blockExplorerUrl: string, txnHash: string): string => {
	//   currently we are not accounting for the case where txnHash is empty or the blockExplorerUrl is empty
	return `${blockExplorerUrl}/tx/${txnHash}`;
};

export const doNothing = () => {
	// do nothing
};

export const bigIntAbs = (n: bigint) => {
	return n < 0n ? -n : n;
};

export function cn(...args: ClassValue[]) {
	return twMerge(clsx(args));
}

export const removeTrailingZeros = (decimalString: string): string => {
	// Check if the input contains a decimal point
	if (decimalString.includes('.')) {
		decimalString = decimalString.replace(/(\.\d*?[1-9])0+$/, '$1'); // Remove trailing zeros after a decimal point
		decimalString = decimalString.replace(/\.0+$/, ''); // Remove trailing decimal part if it's all zeros
	}
	return decimalString;
};
