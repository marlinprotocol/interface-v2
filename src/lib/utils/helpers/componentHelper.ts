import { buttonClasses } from '$lib/atoms/componentClasses';
import { staticImages } from '$lib/components/images/staticImages';
import { addToast } from '$lib/data-stores/toastStore';
import type {
	ButtonModel,
	CommonVariant,
	DividerDirection,
	InputCardVariant,
	TextModel
} from '$lib/types/componentTypes';
import { copyTextToClipboard } from '$lib/utils/helpers/commonHelper';

export const getButtonStyles = (variant: ButtonModel['variant']) => {
	switch (variant) {
		case 'filled':
			return buttonClasses.filled;
		case 'outlined':
			return buttonClasses.outlined;
		case 'greyOutlined':
			return buttonClasses.greyOutlined;
		case 'error':
			return buttonClasses.error;
		case 'text':
			return buttonClasses.text;
		case 'info':
			return buttonClasses.lightblueFilled;
		case 'greyFilled':
			return buttonClasses.greyFilled;
		case 'whiteFilled':
			return buttonClasses.whiteFilled;
		case 'tableConvertButton':
			return buttonClasses.tableConvertButton;
		case 'icon':
			return buttonClasses.icon;
		default:
			return buttonClasses.filled;
	}
};

export const getTextColorClassByVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return 'text-primary';
		case 'secondary':
			return 'text-secondary';
		case 'success':
			return 'text-primary-content';
		case 'info':
			return 'text-primary-content';
		case 'warning':
			return 'text-primary-content';
		case 'error':
			return 'text-error-content';
		case 'disabled':
			return 'text-primary opacity-40';
		case 'grey':
			return 'text-grey-1000';
		default:
			return '';
	}
};

// these color codes are part of the design system, we assume number of rows to be 10 adding more rows would require adding more colors
export const getColorHexForTableRow = (index: number) => {
	const remainder = Math.abs(index) % 10;
	switch (remainder) {
		case 0:
			return '#AF2EFF';
		case 1:
			return '#9868FF';
		case 2:
			return '#4171EA';
		case 3:
			return '#4FA0FF';
		case 4:
			return '#4FC0FF';
		case 5:
			return '#25CFDA';
		case 6:
			return '#19CDB7';
		case 7:
			return '#1EC086';
		case 8:
			return '#3DAE56';
		case 9:
			return '#D8D013';
	}
};

export const getInputCardVariant = (variant: InputCardVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return 'bg-white px-6 py-[30px] rounded-3xl border border-[#D9DADE]';
		case 'primary-mini':
			return 'bg-base-200 py-2 px-4';
		case 'warning':
			return 'bg-red-100 px-4 py-1';
		case 'yellow':
			return 'bg-[#EDBE59]';
		case 'green':
			return 'bg-[#68A843]';
		case 'none':
			return '';
		case 'search':
			return 'bg-grey-200 px-4 py-2.5 rounded-[100px]';
		case 'v2Input':
			return 'relative border border-[#D9DADE] bg-white py-2.5 px-4 rounded-full';
		default:
			return 'bg-base-200 p-4';
	}
};

export const getColorClassByVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'success':
			return 'bg-[#F4F9F0]';
		case 'warning':
			return 'bg-[#FCEFD4]';
		case 'error':
			return 'bg-[#FEE6E6]';
		default:
			return '';
	}
};

export const getIconbyVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'success':
			return staticImages.toastSuccessAlert;
		case 'warning':
			return staticImages.toastWarningAlert;
		case 'error':
			return staticImages.toastErrorAlert;
		default:
			return '';
	}
};

export const getImageClassByVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return 'primary-img';
		case 'info':
		case 'secondary':
			return 'secondary-img';
		case 'warning':
			return 'warning-img';
		case 'error':
			return 'error-img';
		case 'success':
			return 'success-img';
		case 'disabled':
			return 'disabled-img';
		case 'grey':
			return 'grey-img';
		default:
			return 'primary-img';
	}
};

export const getColorHexByVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return '#0F62FE';
		case 'secondary':
			return '#6F6F6F';
		case 'success':
			return '#24A148';
		case 'info':
			return '#0062FF';
		case 'warning':
			return '#f1861b';
		case 'error':
			return '#DA1E28';
		case 'disabled':
			return '#0F62FE';
		case 'grey':
			return '#6F6F6F';
		default:
			return '';
	}
};
export const getColorHexByVariantForTag = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return '#0F62FE23';
		case 'success':
			return '#F4F9F0';
		case 'warning':
			return '#FCEFD4';
		case 'error':
			return '#FEE6E6';
		default:
			return '';
	}
};

// -------------------------------------------- DIVIDER HELPERS --------------------------------------
export const getDividerColor = (variant: DividerDirection) => {
	switch (variant) {
		case 'divider-horizontal':
			return 'bg-black/[0.1]';
		case 'divider-vertical':
			return 'bg-grey-1000';
	}
};

export const getDividerWidth = (variant: DividerDirection) => {
	switch (variant) {
		case 'divider-horizontal':
			return 'w-full';
		case 'divider-vertical':
			return 'w-[1px]';
	}
};

export const getDividerHeight = (variant: DividerDirection) => {
	switch (variant) {
		case 'divider-horizontal':
			return 'h-[1px]';
		case 'divider-vertical':
			return 'h-4';
	}
};

// -------------------------------------------- FONT HELPERS ----------------------------------------

export const getFontSize = (variant: TextModel['variant']) => {
	switch (variant) {
		case 'h1':
			return 'text-4xl';
		case 'h2':
			return 'text-3xl';
		case 'h3':
			return 'text-2xl';
		case 'h4':
			return 'text-xl';
		case 'h5':
			return 'text-lg';
		case 'h6':
		case 'body':
			return 'text-base';
		case 'small':
			return 'text-sm';
		case 'tiny':
			return 'text-xs';
		case 'nav':
			return 'text-sm';
	}
};

export const getDefaultFontWeight = (variant: TextModel['variant'] | undefined) => {
	switch (variant) {
		case 'h1':
		case 'h2':
		case 'h3':
		case 'h4':
		case 'h5':
			return 'font-bold';
		case 'h6':
			return 'font-semibold';
		case 'body':
		case 'small':
		case 'tiny':
			return 'font-normal';
		case 'nav':
			return 'font-semibold';
		default:
			return 'font-normal';
	}
};

// -------------------------------------------- USER INTERACTION HELPERS ----------------------------------------

export const handleCopyClick = (text: string | undefined, successMessage: string) => {
	if (text) {
		copyTextToClipboard(text);
		addToast({
			message: {
				title: 'Copied',
				description: successMessage
			},
			variant: 'success'
		});
	}
};

export const getButtonSize = (size: ButtonModel['size']) => {
	switch (size) {
		case 'tiniest':
			return '';
		case 'tiny':
			return 'h-8';
		case 'smaller':
			return 'h-9';
		case 'small':
			return 'h-10';
		case 'medium':
			return 'h-12';
		case 'large':
			return 'h-14 text-base font-normal';
		default:
			return undefined;
	}
};
