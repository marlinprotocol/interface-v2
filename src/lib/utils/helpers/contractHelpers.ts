import type { Address } from '@web3-onboard/core/dist/types';
import { COMMON_TXN_MESSAGES } from '$lib/utils/constants/messages';
import type { WalletStore } from '$lib/types/storeTypes';
import { addToast } from '$lib/data-stores/toastStore';
import { capitalizeFirstLetter } from '$lib/utils/helpers/commonHelper';
import { ethers } from 'ethers';
import { walletStore } from '$lib/data-stores/walletProviderStore';

let signer: WalletStore['signer'];

walletStore.subscribe((value) => {
	signer = value.signer;
});

export function createSignerContract(contractAddress: Address, contractAbi: any) {
	return new ethers.Contract(contractAddress, contractAbi, signer);
}

export async function createTransaction(
	contractFunctionCall: () => Promise<any>,
	parentFunctionName: string,
	messages?: {
		initiate: { title: string; message: string };
		created: { title: string; message: string };
		success: { title: string; message: string };
		failed: { title: string; message: string };
	}
) {
	try {
		addToast({
			message: {
				title: messages?.initiate.title || COMMON_TXN_MESSAGES.INITIATED.title,
				description: messages?.initiate.message || COMMON_TXN_MESSAGES.INITIATED.message
			},
			variant: 'warning'
		});

		const txn = await contractFunctionCall();

		addToast({
			message: {
				title: messages?.created.title || COMMON_TXN_MESSAGES.CREATED.title,
				description: messages?.created.message || COMMON_TXN_MESSAGES.CREATED.message
			},
			variant: 'warning'
		});

		// waiting for the transaction to be mined
		const approveReciept = await txn.wait();

		// if the transaction is not mined, throw an error and show a toast
		if (!approveReciept) {
			addToast({
				message: {
					title: messages?.failed.title || COMMON_TXN_MESSAGES.FAILED.title,
					description: messages?.failed.message || COMMON_TXN_MESSAGES.FAILED.message
				},
				variant: 'error'
			});
			throw new Error(messages?.failed.message || COMMON_TXN_MESSAGES.FAILED.message);
		}

		// if the transaction is mined, show a toast with success message and return the txn
		addToast({
			message: {
				title: messages?.success.title || COMMON_TXN_MESSAGES.SUCCESS.title,
				description: messages?.success.message
					? COMMON_TXN_MESSAGES.SUCCESS.message + ' ' + messages.success.message
					: COMMON_TXN_MESSAGES.SUCCESS.message
			},
			variant: 'success'
		});
		return { txn: txn, approveReciept: approveReciept };
	} catch (error: any) {
		let title = '';
		let description = error.reason
			? capitalizeFirstLetter(error.reason)
			: COMMON_TXN_MESSAGES.FAILED.message;

		if (error.shortMessage === 'user rejected action') {
			title = 'Transaction Rejected';
			description = 'Transaction rejected by the user';
		}

		addToast({
			message: {
				title: messages?.failed.title || title,
				description
			},
			variant: 'error'
		});
		console.log('error :>> ', error);
		throw new Error(`Transaction Error while creating transaction for ${parentFunctionName}`);
	}
}
