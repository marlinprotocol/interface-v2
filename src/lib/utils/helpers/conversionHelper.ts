import {
	DEFAULT_CURRENCY_DECIMALS,
	DEFAULT_PRECISION,
	SECONDS_IN_HOUR,
	SECONDS_IN_HUNDRED_YEARS
} from '$lib/utils/constants/constants';

import { ethers } from 'ethers';
import { isInputAmountValid } from './commonHelper';

/**
 * Returns duration string for a epoch
 * @param epoch epoch
 * @param mini boolean optional default as false. if true, returns only months, days, hours, mins, secs
 * @param uptoHoursOnly boolean optional default as false. if true, does not return mins and secs
 * @returns string
 * @example 12334422 => 4 months 22 days 18 hours 13 mins 42 secs
 */
export const epochToDurationString = (epoch: number, mini = false) => {
	if (epoch >= SECONDS_IN_HUNDRED_YEARS) return '100+ years';
	const seconds = epoch % 60;
	const minutes = Math.floor(epoch / 60) % 60;
	const hours = Math.floor(epoch / (60 * 60)) % 24;
	const days = Math.floor(epoch / (60 * 60 * 24)) % 30;
	const months = Math.floor(epoch / (60 * 60 * 24 * 30)) % 12;
	const years = Math.floor(epoch / (60 * 60 * 24 * 30 * 12));

	let durationString = '';
	if (years > 0) {
		durationString += years + (years > 1 ? ' years ' : ' year ');
		if (mini) return durationString.trimEnd();
	}
	if (months > 0) {
		durationString += months + (months > 1 ? ' months ' : ' month ');
		if (mini) return durationString.trimEnd();
	}
	if (days > 0) {
		durationString += days + (days > 1 ? ' days ' : ' day ');
		if (mini) return durationString.trimEnd();
	}
	if (hours > 0) {
		durationString += hours + (hours > 1 ? ' hours ' : ' hour ');
		if (mini) return durationString.trimEnd();
	}
	if (minutes > 0) {
		durationString += minutes + (minutes > 1 ? ' mins ' : ' min ');
		if (mini) return durationString.trimEnd();
	}
	if (seconds > 0) {
		durationString += seconds.toFixed() + ' secs';
		if (mini) return durationString;
	}
	return durationString.trimEnd();
};
/**
 * Returns string for a bigInt
 * @param value: bigInt
 * @param bigNumberDecimal: decimal of the token, default set to 18
 * @param precision: number of digits after the decimal point, default set to 2
 * @param commify: default set to true
 * @returns string
 */
export const bigNumberToString = (
	value: bigint,
	bigNumberDecimal = DEFAULT_CURRENCY_DECIMALS,
	precision = DEFAULT_PRECISION,
	commify = true
) => {
	if (value === undefined || value === null) {
		throw new Error('Invalid value');
	}
	if (value === 0n) {
		return '0.00';
	}

	const formattedValue = ethers.formatUnits(value, bigNumberDecimal);
	const [integerPart, decimalPart] = formattedValue.split('.');
	const commifiedIntegerPart = commify ? BigInt(integerPart).toLocaleString() : integerPart;
	const truncatedDecimalPart = decimalPart.slice(0, precision).padEnd(precision, '0');

	return `${commifiedIntegerPart}.${truncatedDecimalPart}`;
};

//return bignumber from string with decimal
export const stringToBigNumber = (value: string, bigNumberDecimal = DEFAULT_CURRENCY_DECIMALS) => {
	if (!value) return 0n;
	if (!isInputAmountValid(value)) return 0n;
	let newValue = value;
	// eslint-disable-next-line prefer-const
	let [integer, fraction] = value.split('.');

	if (fraction && fraction.length > bigNumberDecimal) {
		fraction = fraction.slice(0, bigNumberDecimal);
		newValue = integer + '.' + fraction;
	}
	return ethers.parseUnits(newValue, bigNumberDecimal);
};

export const epochSecToString = (date: number) => {
	return new Date(date * 1000).toLocaleDateString('en-US', {
		year: 'numeric',
		month: 'short',
		day: 'numeric',
		timeZone: 'UTC' // This option specifies the time zone to use
	});
};

export const dateToString = (date: Date) => {
	return date.toLocaleDateString('en-US', {
		year: 'numeric',
		month: 'short',
		day: 'numeric'
	});
};

//short text with ellipsis in the middle
export const shortenText = (text: string, first = 6, last = 4) => {
	if (text.length <= first + last) {
		return text;
	}
	return text.slice(0, first) + '...' + text.slice(text.length - last);
};

export const mPondToPond = (mPond: bigint) => {
	//one mPond is 10^6 pond
	return mPond * BigInt(10) ** BigInt(6);
};

export const pondToMPond = (pond: bigint) => {
	//one pond is 10^-6 mPond
	return pond / BigInt(10) ** BigInt(6);
};

export const convertRateToPerHourString = (
	rate: bigint,
	decimal = DEFAULT_CURRENCY_DECIMALS,
	precision = DEFAULT_PRECISION
) => {
	const rateInHour = rate * BigInt(SECONDS_IN_HOUR);
	return bigNumberToString(rateInHour, decimal, precision);
};

export const convertHourlyRateToSecondlyRate = (rate: bigint) => {
	return rate / BigInt(SECONDS_IN_HOUR);
};
