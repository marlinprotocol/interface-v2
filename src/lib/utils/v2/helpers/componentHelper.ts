import { buttonClasses } from '$lib/atoms/v2/componentClasses';
import { staticImages } from '$lib/components/images/staticImages';
import { addToast } from '$lib/data-stores/toastStore';
import type { ButtonModel, CommonVariant, InputCardVariant } from '$lib/types/v2/componentTypes';
import { copyTextToClipboard } from '$lib/utils/helpers/commonHelper';

export const getButtonStyles = (variant: ButtonModel['variant']) => {
	switch (variant) {
		case 'filled':
			return buttonClasses.filled;
		case 'outlined':
			return buttonClasses.outlined;
		case 'greyOutlined':
			return buttonClasses.greyOutlined;
		case 'error':
			return buttonClasses.error;
		case 'text':
			return buttonClasses.text;
		case 'info':
			return buttonClasses.lightblueFilled;
		case 'greyFilled':
			return buttonClasses.greyFilled;
		case 'whiteFilled':
			return buttonClasses.whiteFilled;
		case 'tableConvertButton':
			return buttonClasses.tableConvertButton;
		default:
			return buttonClasses.filled;
	}
};

export const getInputCardVariant = (variant: InputCardVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return 'bg-white px-6 py-[30px] rounded-3xl border border-[#D9DADE]';
		case 'primary-mini':
			return 'bg-base-200 py-2 px-4';
		case 'warning':
			return 'bg-red-100 px-4 py-1';
		case 'yellow':
			return 'bg-[#EDBE59]';
		case 'green':
			return 'bg-[#68A843]';
		case 'none':
			return '';
		case 'search':
			return 'bg-white px-4 py-2.5 rounded-[100px]';
		case 'v2Input':
			return 'relative border border-[#D9DADE] bg-white py-[20px] px-[18px] rounded-full';
		default:
			return 'bg-base-200 p-4';
	}
};

export const getColorClassByVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'success':
			return 'bg-[#F4F9F0]';
		case 'warning':
			return 'bg-[#FCEFD4]';
		case 'error':
			return 'bg-[#FEE6E6]';
		default:
			return '';
	}
};

export const getIconbyVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'success':
			return staticImages.toastSuccessAlert;
		case 'warning':
			return staticImages.toastWarningAlert;
		case 'error':
			return staticImages.toastErrorAlert;
		default:
			return '';
	}
};

export const getColorHexByVariant = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return '#0F62FE';
		case 'secondary':
			return '#6F6F6F';
		case 'success':
			return '#24A148';
		case 'info':
			return '#0062FF';
		case 'warning':
			return '#f1861b';
		case 'error':
			return '#DA1E28';
		case 'disabled':
			return '#0F62FE';
		case 'grey':
			return '#6F6F6F';
		default:
			return '';
	}
};
export const getColorHexByVariantForTag = (variant: CommonVariant | undefined) => {
	switch (variant) {
		case 'primary':
			return '#0F62FE23';
		case 'success':
			return '#F4F9F0';
		case 'warning':
			return '#FCEFD4';
		case 'error':
			return '#FEE6E6';
		default:
			return '';
	}
};

// -------------------------------------------- USER INTERACTION HELPERS ----------------------------------------
export const handleCopyClick = (text: string | undefined, successMessage: string) => {
	if (text) {
		copyTextToClipboard(text);
		addToast({ message: successMessage, variant: 'success' });
	}
};
