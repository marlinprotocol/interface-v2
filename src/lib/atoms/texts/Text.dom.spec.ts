import TextComponent from './Text.svelte';
import { getDefaultFontWeight, getFontSize } from '$lib/utils/helpers/componentHelper';
import { render, cleanup } from '@testing-library/svelte';
import { describe, it, expect, afterEach } from 'vitest';
import type { TextModel } from '$lib/types/componentTypes';

describe('TextComponent', () => {
	afterEach(() => cleanup());

	const variants = [
		'h1',
		'h2',
		'h3',
		'h4',
		'h5',
		'h6',
		'body',
		'small',
		'tiny',
		'nav'
	] as TextModel['variant'][];

	for (const variant of variants) {
		it(`renders text with correct font size and weight for variant '${variant}'`, () => {
			const testText = `Test ${variant}`;
			const { container, getByTestId } = render(TextComponent, { variant, text: testText });
			expect(container).toMatchSnapshot();
			expect(getByTestId('text').className).toContain(getFontSize(variant));
			expect(getByTestId('text').className).toContain(getDefaultFontWeight(variant));
			expect(getByTestId('text').textContent).toBe(testText);
		});
	}

	it('allows custom styleClass to be applied', () => {
		const styleClass = 'bg-red-100';
		const { getByTestId } = render(TextComponent, { styleClass });
		expect(getByTestId('text').className).toContain(styleClass);
	});

	it('allows custom fontWeight to override default fontWeight', () => {
		const customFontWeight = 'font-semibold';
		const { getByTestId } = render(TextComponent, { fontWeight: customFontWeight });

		expect(getByTestId('text').className).toContain(customFontWeight);
		// ensures default weight is not present
		expect(getByTestId('text').className).not.toContain(getDefaultFontWeight(undefined));
	});

	it('displays text content passed to the component', () => {
		const testText = 'Hello World';
		const { getByTestId } = render(TextComponent, { text: testText });

		expect(getByTestId('text').textContent).toBe(testText);
	});
});
