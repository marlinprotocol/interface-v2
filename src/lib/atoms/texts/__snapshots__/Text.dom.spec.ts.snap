// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextComponent > renders text with correct font size and weight for variant 'body' 1`] = `
<body>
  <div>
    <div
      class=" text-base font-normal"
      data-testid="text"
    >
      Test body
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'h1' 1`] = `
<body>
  <div>
    <div
      class=" text-4xl font-bold"
      data-testid="text"
    >
      Test h1
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'h2' 1`] = `
<body>
  <div>
    <div
      class=" text-3xl font-bold"
      data-testid="text"
    >
      Test h2
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'h3' 1`] = `
<body>
  <div>
    <div
      class=" text-2xl font-bold"
      data-testid="text"
    >
      Test h3
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'h4' 1`] = `
<body>
  <div>
    <div
      class=" text-xl font-bold"
      data-testid="text"
    >
      Test h4
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'h5' 1`] = `
<body>
  <div>
    <div
      class=" text-lg font-bold"
      data-testid="text"
    >
      Test h5
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'h6' 1`] = `
<body>
  <div>
    <div
      class=" text-base font-semibold"
      data-testid="text"
    >
      Test h6
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'nav' 1`] = `
<body>
  <div>
    <div
      class=" text-sm font-semibold"
      data-testid="text"
    >
      Test nav
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'small' 1`] = `
<body>
  <div>
    <div
      class=" text-sm font-normal"
      data-testid="text"
    >
      Test small
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;

exports[`TextComponent > renders text with correct font size and weight for variant 'tiny' 1`] = `
<body>
  <div>
    <div
      class=" text-xs font-normal"
      data-testid="text"
    >
      Test tiny
    </div>
    <!--&lt;Text&gt;-->
  </div>
</body>
`;
